urla5230a98e3_Dhana_Krishnasamy_InMemoryJMS_tgz-pJ8-com_dhana_broker_demo_AppJ8/src/com/dhana/broker/impl/TopicImpl.java

```diff
@@ -0,0 +1,74 @@
+package com.dhana.broker.impl;
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Semaphore;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.dhana.broker.Message;
+import com.dhana.broker.exceptions.DestinationClosedException;
+
+public class TopicImpl extends AbstractDestination implements Topic {
+	static Logger logger = LoggerFactory.getLogger(TopicImpl.class);
+
+	/**
+	 * Distributor is single threaded as the subscriber is not doing much except
+	 * inserting the message in a queue and some scheduling, so In my opinion single thread should
+	 * perform well enough in single JVM (only few subscribers/threads). However
+	 * if it proves to be bottleneck, multiple distribution threads should be
+	 * introduced handing a set of subscribers per thread.
+	 * **/
+	private ExecutorService executor = Executors.newSingleThreadExecutor();
+	private Semaphore permits = new Semaphore(0);
+
+	public TopicImpl(String name) {
+		super(name);
+		executor.execute(new Distributor());
+	}
+
+	public boolean put(Message message) throws DestinationClosedException {
+		boolean ret = super.put(message);
+		permits.release();
+		logger.debug("Message Added:" + message.getMessage());
+		return !ret;
+	}
+
+	class Distributor implements Runnable {
+
+		public void run() {
+			while (true) {
+				try {
+					logger.debug("Waiting for new message");
+					permits.acquire();
+					Message msg = queue.poll();
+					if (msg == null) {
+						logger.error("Received null message from the queue, programming error??? Ignoring the message");
+						break;
+					}
+					if (msg == poison) {
+						logger.info("Received close signal, closing queue:"
+								+ name);
+						break;
+					}
+					// Bottleneck prone - Intro multiple thread per set of subscribers
+					for (AsyncMessageConsumer subscriber : subscribers) {
+						subscriber.onMessage(msg.clone());
+						logger.debug("Message sent to a client");
+					}
+				} catch (InterruptedException e) {
+					Thread.currentThread().interrupt();
+					break;
+				}
+			}
+
+		}
+
+	}
+
+	public void close() {
+		super.close();
+		executor.shutdownNow();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Call `shutdown()` to reject any incoming tasks and `awaitTermination()` to wait for existing tasks to complete or until a timeout is reached. To stop the execution and processing of tasks, if necessary, call `shutdownNow()`. Check the return value of `awaitTermination()` and handle when the method times out while waiting for other threads to stop executing. See the following diffs on GitHub that make this code change: [diff1](https://github.com/glowroot/glowroot/commit/81de5a3699c59e0cf98d485c380c1034e6cd28a5#diff-c608d2c40790b0e0e5228e6647f69729caa756bbe9127169139406314b80897eL143), [diff2](https://github.com/apache/directory-kerby/commit/3ed0e7cd3b8a0c3ce29293153e4e72f422f0637f#diff-eb42f072849f520563ec2a686789a836195f18075cc222ef355c6ea710b0a360L84)

Learn more: [ExecutorService in the Java Platform, Standard Edition 8 API Specification](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ExecutorService.html)